File: src/com/vidyasource/strategy/java/StrategyExample.java
package com.vidyasource.strategy.java;

class StrategyExample {
    public static void main(String[] args) {
        Strategy addStrategy = new AddStrategy();
        Strategy subtractStrategy = new SubtractStrategy();
        Strategy multiplyStrategy = new MultiplyStrategy();

        System.out.println("Add:      " + execute(addStrategy, 3, 4));
        System.out.println("Subtract: " + execute(subtractStrategy, 3, 4));
        System.out.println("Multiply: " + execute(multiplyStrategy, 3, 4));
    }

    private static int execute(Strategy strategy, int x, int y) {
        return strategy.apply(x, y);
    }
}

File: src/com/vidyasource/strategy/scala/StrategyExample.scala
package com.vidyasource.strategy.scala

object StrategyExample extends App {
  val addStrategy: Function2[Int, Int, Int] = (a: Int, b: Int) => a + b
  val subtractStrategy: Function2[Int, Int, Int] = (a: Int, b: Int) => a - b
  val multiplyStrategy: Function2[Int, Int, Int] = (a: Int, b: Int) => a * b

  def execute(callback: Function2[Int, Int, Int])(x: Int)(y: Int): Int = callback.apply(x, y)

  println(s"Add:      ${execute(addStrategy)(3)(4)}")
  println(s"Subtract: ${execute(subtractStrategy)(3)(4)}")
  println(s"Multiply: ${execute(multiplyStrategy)(3)(4)}")
}

